canopen/node/local.py

```diff
@@ -11,16 +13,35 @@ def __init__(self, node_id, object_dictionary):
         self.callbacks = []
 
         self.sdo = SdoServer(0x600 + self.id, 0x580 + self.id, self)
+        self.nmt = NmtSlave(self.id, self)
+        self.add_callback(self._producer_hearbeat_time_callback)
 
     def associate_network(self, network):
         self.network = network
         self.sdo.network = network
+        self.nmt.network = network
         network.subscribe(self.sdo.rx_cobid, self.sdo.on_request)
+        network.subscribe(0, self.nmt.on_command)
 
     def remove_network(self):
         self.network.unsubscribe(self.sdo.rx_cobid)
+        self.network.unsubscribe(0)
         self.network = None
         self.sdo.network = None
+        self.nmt.network = None
 
     def add_callback(self, callback):
         self.callbacks.append(callback)
+
+    def _producer_hearbeat_time_callback(self, **kwargs):
+        if kwargs["index"] == 0x1017:
+            if kwargs["data"] is None:
+                # This is a read callback. If we return none
+                # the data will be read from local storage
+                return None
+            elif kwargs["data"] == 0:
+                self.nmt.stop_heartbeat()
+            else:
+                (hearbeat_time, ) = struct.unpack_from("<H", kwargs["data"])
+                self.nmt.start_heartbeat(hearbeat_time)
+            return 0x0201
```

Comment:
Is this necessary?

Reply:
I assume you mean the return value?

I'm not sure what to actually return. The server that calls the callback looks like this:

        for callback in self._callbacks:
            status = callback(index=index, subindex=subindex, od=obj, data=data)
            if status:
                break

So I guess the callback that handles the index, subindex should return something other than None to indicate that the download was handled.

Any suggestion ?

Reply:
I think the idea was that the function could return `True` if no other callbacks should be called for this write operation. Not sure if this is actually necessary though.

Reply:
Fixed. Return True to indicate that the callback has been handled.

Maybe a future improvements could be to have seperate upload and download callback. Then it wouldn't be necessary in the callback to deduce if the callback is called due to a upload or a download. 

For a download callback we could return True/False depending of the callback was handled. For the upload callback we could return the value or none.